### PR TITLE
test: add a test for watchdog timers

### PR DESCRIPTION
**Pull request by dsseng**
# Pull Request

## What? (description)
Add a test for watchdog

## Why? (reasoning)
Ensure integration tests cover all code

## Acceptance

Please use the following checklist:

- [ ] you linked an issue (if applicable)
- [x] you included tests (if applicable)
- [ ] you ran conformance (`make conformance`)
- [x] you formatted your code (`make fmt`)
- [x] you linted your code (`make lint`)
- [ ] you generated documentation (`make docs`)
- [ ] you ran unit-tests (`make unit-tests`)

> See `make help` for a description of the available targets.

### CODE REVIEW

**Comment by smira** [comment]
I'm going to take a look into this to add more generic helpers to help with patching machine configuration.

**Comment by smira** [comment]
Feedback/notes:

* we should use `test:` for changes around tests 
* I added some base helpers to help with patching - the easiest is to use strategic merge config patches to update/add multi-doc machine config documents.
* it's always better to use native types than to build YAML as bytes manually
* after applying machine config we can't expect immediate change being applied, as controllers work async - so we use controller output resources as a signal that the word had been done
* extract common helper to read watchdog properties

**Comment by smira** [comment]
```shell
$ _out/integration-test-linux-amd64 -test.v -talos.crashdump=false -talos.provisioner=qemu -talos.talosctlpath=$PWD/_out/talosctl-linux-amd64 -test.run TestIntegration/api.Watch
=== RUN   TestIntegration
=== RUN   TestIntegration/api.WatchdogSuite
=== RUN   TestIntegration/api.WatchdogSuite/TestWatchdogSysfs
    watchdog.go:69: testing watchdog on node 172.20.0.2
    api.go:593: patched machine config: Applied configuration without a reboot
    api.go:593: patched machine config: Applied configuration without a reboot
--- PASS: TestIntegration (0.30s)
    --- PASS: TestIntegration/api.WatchdogSuite (0.29s)
        --- PASS: TestIntegration/api.WatchdogSuite/TestWatchdogSysfs (0.14s)
PASS
```

File: internal/integration/api/watchdog.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+	suite.T().Logf("testing watchdog on node %s", node)
+
+	// build a Talos API context which is tied to the node
+	nodeCtx := client.WithNode(suite.ctx, node)
+
+	// pick a watchdog
+	const watchdog = "watchdog0"
+
+	wdState := suite.readWatchdogSysfs(nodeCtx, watchdog, "state")
+	suite.Require().Equal("inactive", wdState)
```
Comment (dsseng) [inline internal/integration/api/watchdog.go:79]: What if the test node has WDT already enabled? Maybe we should first remove the document and assert inactive state?

**Comment by dsseng** [comment]
Thanks a lot for your help, will see and do anything else to be done if anything tomorrow

File: internal/integration/api/watchdog.go (hunk 1 above)
Comment (smira) [reply internal/integration/api/watchdog.go:79]: yes, we can add a remove on top, but as the test removes the watchdog, we might assume that a "clean" state is that watchdog is not active

we assume that tests run in a clean environment

Comment (smira) [reply internal/integration/api/watchdog.go:79]: either way is fine

File: internal/integration/api/watchdog.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+	wdState = suite.readWatchdogSysfs(nodeCtx, watchdog, "state")
+	suite.Require().Equal("active", wdState)
+
+	wdTimeout := suite.readWatchdogSysfs(nodeCtx, watchdog, "timeout")
+	suite.Require().Equal("120", wdTimeout)
+
+	cfgDocument.WatchdogTimeout = 60 * time.Second
+	suite.PatchMachineConfig(nodeCtx, cfgDocument)
+
+	_, err = suite.Client.COSI.WatchFor(nodeCtx, runtimeres.NewWatchdogTimerStatus(runtimeres.WatchdogTimerConfigID).Metadata(), state.WithEventTypes(state.Created, state.Updated))
```
Comment (smira) [inline internal/integration/api/watchdog.go:102]: It won't work this way, as the resource is already present, so you want to watch for a resource to have expected state, that is timeout of 60 seconds

Comment (dsseng) [reply internal/integration/api/watchdog.go:102]: isn't state.Updated expected to catch this?

Comment (smira) [reply internal/integration/api/watchdog.go:102]: first of all, you have both there; second it's always a race - did you start watching before it got updated or after, you never know; does any other controller update the resource (e.g. put a finalizer on it)

what you need here is: https://pkg.go.dev/github.com/cosi-project/runtime@v0.4.1/pkg/state#WithCondition

```go
state.WithCondition(func (r resource.Resource) bool {
   return r.(*runtime.WatchdogTimeStatus).TypedSpec().Timeout == 2 * time.Minute
})
```

Comment (smira) [reply internal/integration/api/watchdog.go:102]: (wrong timeout, but I hope my point is clear)

File: internal/integration/api/watchdog.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+	suite.Require().NoError(err)
+
+	wdState = suite.readWatchdogSysfs(nodeCtx, watchdog, "state")
+	suite.Require().Equal("active", wdState)
+
+	state.WithCondition(func(r resource.Resource) (bool, error) {
```
Comment (smira) [inline internal/integration/api/watchdog.go:109]: this statement has no effect, it should instead be moved into line 103 as the argument to the `WatchFor` function

File: internal/integration/api/watchdog.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+		return (r.(*runtimeres.WatchdogTimerStatus).TypedSpec().Timeout == 1*time.Minute), nil
+	})
+	wdTimeout = suite.readWatchdogSysfs(nodeCtx, watchdog, "timeout")
+	suite.Require().Equal("60", wdTimeout)
+
+	cfgDocument.WatchdogTimeout = 90 * time.Second
+	suite.PatchMachineConfig(nodeCtx, cfgDocument)
+
+	state.WithCondition(func(r resource.Resource) (bool, error) {
```
Comment (smira) [inline internal/integration/api/watchdog.go:118]: same here

**Review by smira — APPROVED** [review]
fixed up small issues to get this PR merged

**Pull request merged**